### PR TITLE
New custom state output syntax and runtime API

### PR DESCRIPTION
**Pull request by tomrav**
- Stylable now outputs custom states using class names instead of `data-*` attributes
- Stylable runtime stylesheet function changed to a new API, see this [comment](https://github.com/wix/stylable/issues/486#issuecomment-461545325) for further info
- The class name optimizer has been instructed to ignore class names representing states (as those are not mapped during runtime at the moment)
- States target output now uses a constant throughout the code
- Testing utilities have been refactored to the new runtime API, and class based states
- Refactor `react-scripts` components to utilize the new runtime API
- Create new testing `dom-test-util` function - `getStateValueFromClassName`
- Remove deprecated "cssStates" function from the stylable transformer

### CODE REVIEW

File: packages/core/src/pseudo-states.ts [hunk 1]
```diff
@@ -37,7 +36,7 @@ export function processPseudoStates(value: string, decl: postcss.Declaration, di
     statesSplitByComma.forEach((workingState: ParsedValue[]) => {
         const [stateDefinition, ...stateDefault] = workingState;
 
-        if (isCSSVarProp(stateDefinition.value.trim())) {
+        if (stateDefinition.value.trim().startsWith('-')) {
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/pseudo-states.ts]: also startsWith(delimiter)

File: packages/dom-test-kit/src/stylable-dom-util.ts [hunk 2]
```diff
@@ -28,36 +30,64 @@ export class StylableDOMUtil {
...
+        if (!element.className.includes(stateName)) {
+            return null;
+        }
+
+        const booleanState = this.style.$cssStates({ [stateName]: true });
+        if (element.classList.contains(booleanState)) {
+            return true;
+        }
+
+        const baseState = this.style.$cssStates({ [stateName]: '_' }).slice(0, -3);
```
Comment (barak007) [inline packages/dom-test-kit/src/stylable-dom-util.ts]: please give the _ a name like singleCharStateName

File: packages/dom-test-kit/src/stylable-dom-util.ts [hunk 3]
```diff
@@ -28,36 +30,64 @@ export class StylableDOMUtil {
...
+
+        let paramValue = '';
+        element.classList.forEach(cls => {
+            if (!paramValue && cls.startsWith(baseState)) {
+                const param = cls.slice(baseState.length);
```
Comment (barak007) [inline packages/dom-test-kit/src/stylable-dom-util.ts]: I think we want the utils like getStateClassNameValuePart

File: packages/optimizer/src/classname-optimizer.ts [hunk 4]
```diff
@@ -8,24 +8,32 @@ export class StylableClassNameOptimizer {
...
         };
     }
-    public rewriteSelector(selector: string) {
+    public rewriteSelector(selector: string, namespace: string) {
         const ast = parseSelector(selector);
         traverseNode(ast, node => {
             if (node.type === 'class') {
-                if (!this.context.names[node.name]) {
-                    this.generateName(node.name);
+                if (!node.name.startsWith(`${namespace}__`)) {
```
Comment (barak007) [inline packages/optimizer/src/classname-optimizer.ts]: import the '__' from...

File: packages/runtime/src/css-runtime-stylesheet.ts [hunk 5]
```diff
@@ -66,59 +60,70 @@ export function create(
...
+        }
+
+        if (stateValue === true) { // boolean state
+            return autoStateClassName(stateName, namespace, false);
+        }
+
+        const valueAsString = stateValue.toString();
+
+        // tslint:disable-next-line: max-line-length
+        return `${autoStateClassName(stateName, namespace, true)}${valueAsString.length}_${valueAsString.replace(/\s/gm, '_')}`;
```
Comment (barak007) [inline packages/runtime/src/css-runtime-stylesheet.ts]: use helper function to generate this string

File: packages/core/tests/pseudo-states.spec.ts [hunk 6]
```diff
@@ -469,55 +469,6 @@ describe('pseudo-states', () => {
         });
 
         describe('boolean', () => {
-
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/core/tests/pseudo-states.spec.ts]: Was this comment left for something specific, or?

File: packages/core/tests/pseudo-states.spec.ts [hunk 7]
```diff
@@ -615,13 +566,13 @@ describe('pseudo-states', () => {
 
                 expect(res.meta.diagnostics.reports, 'no diagnostics reported for native states').to.eql([]);
                 expect(res).to.have.styleRules({
-                    1: '.entry--my-class[data-entry-state1="someString"] {}'
+                    1: '.entry--my-class.entry___state110_someString {}'
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/core/tests/pseudo-states.spec.ts]: consider fast. :)

File: packages/core/tests/pseudo-states.spec.ts [hunk 8]
```diff
@@ -661,7 +612,7 @@ describe('pseudo-states', () => {
 
                     expect(res.meta.diagnostics.reports, 'no diagnostics reported for native states').to.eql([]);
                     expect(res).to.have.styleRules({
-                        1: '.entry--my-class[data-entry-statewithdefault="myDefault String"] {}'
+                        1: '.entry--my-class[class~="entry___stateWithDefault16_myDefault_String"] {}'
```
[the remark this answers is not in the document]
Comment (tomrav) [reply packages/core/tests/pseudo-states.spec.ts]: Not sure I get what you mean here.

File: packages/core/src/pseudo-states.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/pseudo-states.ts]: not sure about the trim I think you can remove it

File: packages/core/src/pseudo-states.ts [hunk 9]
```diff
@@ -22,7 +21,7 @@ export const stateErrors = {
     TOO_MANY_STATE_TYPES: (name: string, types: string[]) => `pseudo-state "${name}(${types.join(', ')})" definition must be of a single type`,
     NO_STATE_TYPE_GIVEN: (name: string) => `pseudo-state "${name}" expected a definition of a single type, but received none`,
     TOO_MANY_ARGS_IN_VALIDATOR: (name: string, validator: string, args: string[]) => `pseudo-state "${name}" expected "${validator}" validator to receive a single argument, but it received "${args.join(', ')}"`,
-    STATE_VARIABLE_NAME_CLASH: (name: string) => `state "${name}" definition cannot begin with "--"`
+    STATE_VARIABLE_NAME_CLASH: (name: string) => `state "${name}" declaration cannot begin with a "-" chararcter`
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/core/src/pseudo-states.ts]:
```suggestion
    INVALID_STATE_NAME: (name: string) => `state "${name}" declaration cannot begin with a "-" chararcter`
```

**Pull request merged**